diffrax/_progress_bar.py

```diff
@@ -0,0 +1,174 @@
+import abc
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+from jax.experimental import io_callback
+from jaxtyping import Array, PyTree
+
+from ._custom_types import FloatScalarLike, IntScalarLike
+
+
+class AbstractProgressBar(eqx.Module):
+    @abc.abstractmethod
+    def init(self) -> PyTree[Array]:
+        """Creates a new progress bar and returns its unique identifier"""
+
+    @abc.abstractmethod
+    def step(self, state: PyTree[Array], progress: FloatScalarLike) -> PyTree[Array]:
+        """Updates the progress bar and returns the last printed progress value"""
+
+    @abc.abstractmethod
+    def close(self, state: PyTree[Array]):
+        """Closes the progress bar"""
+
+
+class NoProgressBar(AbstractProgressBar):
+    def init(self) -> None:
+        return None
+
+    def step(self, state, progress: FloatScalarLike) -> None:
+        return state
+
+    def close(self, state):
+        pass
+
+
+class _TextProgressBarState(eqx.Module):
+    progress: FloatScalarLike
+
+    def __init__(self, progress):
+        self.progress = progress
```

Comment:
This is superfluous by the way: Equinox modules are dataclasses so a default `__init__` method will be synthesise automatically from the fields.